### PR TITLE
feat(physics): add delta param to move_and_slide

### LINKED ISSUE #24322: Add an optional `delta` parameter to `KinematicBody.move_and_slide()`
Let's assume that we have a client (running at 60 FPS) that is sending inputs to the server each frame. The server (running at 30 FPS) collects these inputs and tries to return the correct position to the client doing the physics calculation on his side. If we try using move_and_slide for this purpose we got an incorrect position because of the difference in the server and client FPS. move_and_slide function should have exposed a delta time parameter, so we can use delta obtained from client.

### PR DESCRIPTION
**Pull request by jonbonazza**
This adds an optional delta parameter to KinematicBody3D::move_and_slide,
KinematicBody3D::move_and_slide_with_snap,
KinematicBody2D::move_and_slide, and
KinematicBody2D::move_and_slide_with_snap that can be used to override
the delta value used in physics calculations.

fixes #24322

### CODE REVIEW

File: scene/2d/physics_body_2d.cpp
Comment (bojidar-bg) [inline scene/2d/physics_body_2d.cpp]: Comment should be moved as well. (likewise for 3D)

**Comment by Calinou** [comment]
Would it make sense to have this for `move_and_collide()` as well?

**Comment by bojidar-bg** [comment]
`move_and_collide` takes the distance, not the velocity.
For RigidBody, #25068 might be better.

**Comment by jonbonazza** [comment]
@Calinou yea I didn't add it to `move_and_collide` because of the reason that @bojidar-bg pointed out.

Should I also implement this for `RigidBody` or should a separate PR be made for #25068 instead?

**Comment by jonbonazza** [comment]
I rebased onto latest master and solved the conflict that was created with commit 750f343e4a461c588494bf779b9be7c1be8ebc68.

I also moved the comments to a better spot. Will this suffice @bojidar-bg ?

File: scene/2d/physics_body_2d.cpp
Comment (bojidar-bg) [inline scene/2d/physics_body_2d.cpp]: You could reuse the parameter here, especially since it is not `const float&`.

File: scene/2d/physics_body_2d.cpp
Comment (bojidar-bg) [inline scene/2d/physics_body_2d.cpp]: In most other cases, the default value for "automatically computed" is `-1.0`, not `0.0`. (It might also be useful to change the `if` to `delta < 0`, so that if someone messes up delta computations and gets an actual zero, they don't suddenly get the real delta instead)

File: doc/classes/KinematicBody.xml
Comment (bojidar-bg) [inline doc/classes/KinematicBody.xml]: There shouldn't be a comma between `useful` and `for`.

**Review by bojidar-bg — APPROVED** [review]
Changes look good to me overall.

**Comment by jonbonazza** [comment]
@bojidar-bg I'm in the middle of a move currently and won't have my dev machine that has godot's build env setup for another few weeks. I'll have to come back and resolve these last few changes after that.

**Comment by AndreaCatania** [comment]
Any news about this one?

**Comment by jonbonazza** [comment]
Im planning to resolve the remaining comments today

**Comment by jonbonazza** [comment]
@bojidar-bg @AndreaCatania sorry I took so long to get around to this, but I finally made the minor updates that were requested. Also rebased against master. let me know if there's anything else. :)

File: scene/3d/physics_body.cpp
Comment (AndreaCatania) [inline scene/3d/physics_body.cpp]: Can you please use `<=` we have to avoid the possibility to send 0 as delta to avoid undefined behaviour.

**Review by AndreaCatania — CHANGES_REQUESTED** [review]

File: scene/3d/physics_body.cpp
Comment (jonbonazza) [reply scene/3d/physics_body.cpp]: I originally had that, but was asked to change it by @bojidar-bg in a previous comment. I agree with you though, it should probably protect against 0. I will change it back.

Comment (jonbonazza) [reply scene/3d/physics_body.cpp]: @AndreaCatania I've made the change you have requested. Let me know if there is anything else. :)

**Review by AndreaCatania — APPROVED** [review]

File: scene/3d/physics_body.cpp
Comment (AndreaCatania) [reply scene/3d/physics_body.cpp]: Perfect, thanks! I've approved it.

**Comment by jonbonazza** [comment]
I've rebased onto master. Can we get this merged so that I don't need to keep maintaining it through all the sweeping changes going on in master atm? @Calinou @akien-mga

**Comment by jonbonazza** [comment]
@akien-mga @Calinou ping

**Comment by Calinou** [comment]
I'm not allowed to merge pull requests on the main Godot repository, sorry. You'll have to wait for Akien to approve it.

**Comment by jonbonazza** [comment]
@akien-mga senpaiii 🙏

File: doc/classes/KinematicBody3D.xml
Comment (AndreaCatania) [inline doc/classes/KinematicBody3D.xml:116]: Typo, Fixing this should also fix the CI static check.

**Comment by jonbonazza** [comment]
@akien-mga any update on this one?

File: doc/classes/KinematicBody3D.xml
Comment (Calinou) [inline doc/classes/KinematicBody3D.xml:142]: Documentation for the `delta` parameter should be added here as well.

Comment (jonbonazza) [reply doc/classes/KinematicBody3D.xml:142]: @Calinou I'm not sure I understand. The last line of the description does include details about the `delta` parameter.

Comment (jonbonazza) [reply doc/classes/KinematicBody3D.xml:142]: @Calinou I purposely didnt add it here because it says "similar to method `move_and_slide`" and doesn't detail any of the other parameters that `move_and_slide` has. It just explains the addition of the snap bit.

**Comment by jonbonazza** [comment]
@akien-mga just rebased again... can we please get this merged? 🙏

**Comment by madmiraal** [comment]
Maybe I'm missing something, but I don't understand why this change is needed; especially in the context of "This is useful for synchronizing physics calculations across multiple devices (e.g. a server and clients)."

Changing `delta` only changes this body's perceived velocity, with the additional negative side-effect of changing the detected floor velocity this body is standing on; causing it to slip.

**Comment by AndreaCatania** [comment]
It's useful when two steps are combined.
- Client executes two steps with `Velocity = 1`, `Delta = 1`
- The server combines those, so `Velocity = 1`, `Delta = 2`

True is that the velocity may change, instead of the delta. But change the delta make the code look right: -> Two steps -> double delta -> same velocity.

**Comment by madmiraal** [comment]
> It's useful when two steps are combined.
>
>    Client executes two steps with Velocity = 1, Delta = 1
>    The server combines those, so Velocity = 1, Delta = 2

These are already the deltas that would be used. They are calculated from the `Physics Fps` setting. They do not need to be passed into the `move_and_slide()` function.

**Comment by jonbonazza** [comment]
@madmiraal the point of this pr is to override that setting. 

For instance a server and client might not have the same physics fps. When a client performs client side prediction and then reconciles the server state, it's important that the physics fps matches on both sides. 

This doesn't require everyone to pass it in. If you want the normal behavior just dont pass in an override--it defaults to the normal behavior.

**Comment by madmiraal** [comment]
@jonbonazza In a client-server scenario, either the server is the master and the clients update their positions based on the server's calculations, or the server and clients must have the same fps (even then it is hard to ensure they stay in sync). Either way there is no need to specify the delta in the `move_and_slide()` function.

As stated previously: changing `delta` only changes this body's perceived velocity, with the additional negative side-effect of changing the detected floor velocity this body is standing on; causing it to slip.

**Comment by jonbonazza** [comment]
@madmiraal you still aren't understanding and im tired of trying to explain. I don't know why you are do against this. If you don't feel like you need it then don't use it.

**Comment by Calinou** [comment]
There are cases where you want the server to use a 'ower physics FPS than the client. This is common in MMOs to save CPU resources while keeping the client-side prediction smooth and reactive.

**Comment by bojidar-bg** [comment]
Let me add another case in which I think having a custom delta is useful: tick-based simulation.

When implementing a fully-synchronized peer-to-peer networking protocol, similar to the ones found in certain RTS games, one wants to collect inputs from all players, then advance the simulation by one tick. When the internet conditions are near-optimal, you can gather inputs for, say, 4 frames in the future, and advance on a fixed timestep. Otherwise, if you cannot gather inputs for a certain frame in time, the simulation has to be stopped until the inputs arrive.

Now, it is certainly possible to use `_physics_process` along with pausing the SceneTree, but that is hacky (you might want to have actual physics simulation going on in order to animate stuff like bones, building rubble, etc.) and limited (you might have to pause a frame earlier when you detect that data is missing, not very sure about this).

Therefore, rolling a custom system for "ticks" is not a bad idea overall. The trouble with it is that one currently loses the opportunity to use KinematicBody's `move_and_slide`/`move_and_slide_with_snap`, since those require  autodetected `delta`.

---

Another application off the top of my head would be to set `delta` to `1` in order to use `move_and_slide` with a given change in position directly, instead of having to reimplement the sliding functionality using `move_and_collide`.

---

If there are problems related to floor velocity when the delta is overridden, those should be pointed out and a fix should be discussed. Personally, I'm unable to spot such an issue myself when looking over the code (especially considering that it wouldn't happen when using it the current `move_and_slide` with `_process`, right?).
Otherwise, I personally don't see a reason to delay merging this PR over and over again.

**Comment by madmiraal** [comment]
> I don't know why you are do against this.

@jonbonazza I refer you to the [Best practices for engine contributors](https://docs.godotengine.org/en/stable/community/contributing/best_practices_for_engine_contributors.html). The change must solve a problem that exists, and I don't see the problem this change is trying to solve.

> There are cases where you want the server to use a lower physics FPS than the client. This is common in MMOs to save server CPU resources while keeping the client-side prediction smooth and reactive.

The server and the client already can have different physics fps. This PR is not needed for that.

> When implementing a fully-synchronized peer-to-peer networking protocol, similar to the ones found in certain RTS games, one wants to collect inputs from all players, then advance the simulation by one tick...

This is certainly a nice idea, but this PR doesn't achieve that. This is the scenario that requires the server and clients to have the same fixed fps, and it's very hard to ensure they stay in sync. Many games that implement this approach have the dreaded "Desync" crash. It crashes because it's unrecoverable. Having a different delta would instantly create a desync.

> Another application off the top of my head would be to set delta to 1 in order to use move_and_slide with a given change in position directly, instead of having to reimplement the sliding functionality using move_and_collide.

I'm not sure what you're suggesting, but either you're specifying the position or you want `move_and_slide()` to calculate the position for you using the correct delta that it retrieves.

> I personally don't see a reason to delay merging this PR over and over again.

I do not see the need for this change. If there is no need, I think it should not be merged.

**Comment by akien-mga** [comment]
For this kind of advanced use case, wouldn't it be better to use `move_and_collide()` to which you can already pass a custom delta?
`move_and_slide()` is just a higher level convenience method that calls `move_and_collide()` with the velocity and internal delta, and handles the potential sliding on non-blocking collisions. That's easy to reimplement in user code, as it's what Godot 2.x used to have as only `move()` API:

https://docs.godotengine.org/en/2.1/learning/features/physics/kinematic_character_2d.html#solution

I'm not saying adding this parameter to `move_and_slide` is a no-go, but I agree with @madmiraal that it needs a well-documented and consensual use case. There should be a proposal on [godot-proposals](https://github.com/godotengine/godot-proposals) that describes the use case, potential workarounds, etc., so that core developers can assess whether this is a good change to the API.

**Comment by jonbonazza** [comment]
Jfc, we've provided multiple well documented use cases and everyone who is familiar with server authoritative multiplayer and the strategies involved have been saying that this is a useful PR. The people who are not seeing the value of this clearly have no experience with server authoritative multiplayer and its strategies.

**Comment by jonbonazza** [comment]
> @jonbonazza I refer you to the [Best practices for engine contributors](https://docs.godotengine.org/en/stable/community/contributing/best_practices_for_engine_contributors.html). The change must solve a problem that exists, and I don't see the problem this change is trying to solve.

@madmiraal there are multiple people in here who _do_ see the value. Just because _you_ don't donesn't mean it doesn't exist.

> The server and the client already can have different physics fps. This PR is not needed for that.

Yes it does, because in order for this to be possible, strategies like client-side prediciton, server reconciliation, and entity interpolation all need to be employed to gaurantee a smooth experience and _these_ are the strategies that require this change. I'm not going to go into the details of how these strategies are implemented, but just trust us that this change is a requirement for this. 

I should note that this isn't a novel idea. This is literally how ever modern MMO works, so it's not like I am just pulling this idea out of my butt.

**Comment by akien-mga** [comment]
@jonbonazza I can appreciate that you might be frustrated that this PR is being challenged, but reaching a consensus on the feature is part of the normal workflow.

Please keep the discussion polite and avoid terms like "Jfc" or "pulling this idea out of [your] butt".

See https://godotengine.org/code-of-conduct

**Comment by jonbonazza** [comment]
> For this kind of advanced use case, wouldn't it be better to use `move_and_collide()` to which you can already pass a custom delta?
> `move_and_slide()` is just a higher level convenience method that calls `move_and_collide()` with the velocity and internal delta, and handles the potential sliding on non-blocking collisions. That's easy to reimplement in user code, as it's what Godot 2.x used to have as only `move()` API:
> 
> https://docs.godotengine.org/en/2.1/learning/features/physics/kinematic_character_2d.html#solution
> 
> I'm not saying adding this parameter to `move_and_slide` is a no-go, but I agree with @madmiraal that it needs a well-documented and consensual use case. There should be a proposal on [godot-proposals](https://github.com/godotengine/godot-proposals) that describes the use case, potential workarounds, etc., so that core developers can assess whether this is a good change to the API.
> 
> One way to demonstrate the validity of this change could be to make a demo project using the feature for something that cannot be done in the engine currently.

> @jonbonazza I can appreciate that you might be frustrated that this PR is being challenged, but reaching a consensus on the feature is part of the normal workflow.
> 
> Please keep the discussion polite and avoid terms like "Jfc" or "pulling this idea out of [your] butt".
> 
> See https://godotengine.org/code-of-conduct

@akien-mga  I'm sorry for being frustrated, but I'm a truly a bit annoyed here. This PR has existed for months now, long before the godot-proposals existed, it has been reviewed and approved several times by many parties with actual knowledge of how this stuff works in the industry. I have been trying to get you to come in here and take a look for _months_ and it's been radio silence. Now, all of a sudden. One person comes in here and starts to bash on it due to ignorance, and you finally show up just to tell me I need a proposal and all this jazz? Like, I get it, a proper process should be followed, but this PR predates that process, and I have been begrudgingly maintaining this PR and fixing conflicts and rebasing many many times. I hope you can understand my frustrations here.

> @jonbonazza I can appreciate that you might be frustrated that this PR is being challenged, but reaching a consensus on the feature is part of the normal workflow.
> 
> Please keep the discussion polite and avoid terms like "Jfc" or "pulling this idea out of [your] butt".
> 
> See https://godotengine.org/code-of-conduct

**Comment by clayjohn** [comment]
Hi @jonbonazza I understand the process can be frustrating at times. It can also be very annoying to have to keep rebasing PRs when you don't know when they will be merged. The reason for delay for many enhancements, including this one, is a lack of full understanding (on our part) of the necessity of the change. 

While you and the people you know have expressed a desire to have this change in the engine, it is difficult from our perspective as reviewers to understand whether this is functionality that is needed by the engine as a whole. Reviewers are typically generalists not subject-matter experts, so you need to prove to us that the change is important and needed. At the end of the day, we are the ones responsible to maintain the code you add, so we like to ensure that it is code that is desirable for a large proportion of people.

This is why we introduced the Godot Improvements Proposal ("GIP") process. In September 2019 we started asking people to submit proposals through a GIP before making PRs. I can see that you missed the memo on that as you submitted this PR in November of 2019 without having gone through the proposal process. The GIP process allows the community and core devs to evaluate a new feature based on its use case, and the problems it intends to solve. If a PR is opened directly, we end up evaluating the PR on the merits of its implementation rather than its desirability. 

The situation we are in now is exactly what we intended to avoid with the GIP process. Your PR has cycled through a few reviews and has had to be rebased multiple times because no one settled the desirability of the feature before you made the PR. Of course now there has been lots of discussion, and it appears that many contributors are in favour of your changes. However, that doesn't excuse you from justifying the change to others. @akien-mga has tried to assist you in ironing out the last possible objections to your PR. 

All you need to do is clarify precisely the use case for this change and why it can't be implemented through script (i.e. by using ``move_and_collide`` as Akien suggested). Doing so helps us justify adding complexity to the engine that we will have to deal with down the road. 

Thanks in advance. :)

**Comment by jonbonazza** [comment]
Hi @clayjohn the problem here is that the necessity for this change is very specific to the normal implementation of server authoritative multiplayer, _specifically_ when the server and client physics process intervals differ (This is not always the case, but sometimes it is depending on the game's design. As an example, FFXI Online takes this approach). Now, explaining all of the implementation details and nuances of server authoritative multiplayer would be far too large of a task to do here. The best I can do is to point you to the following series of articles by the famous Gabriel Gambetta which do a great job of illustrating all of the facets of implementing this kind of multiplayer experience: 

1. https://www.gabrielgambetta.com/client-server-game-architecture.html 
2. https://www.gabrielgambetta.com/client-side-prediction-server-reconciliation.html
3. https://www.gabrielgambetta.com/entity-interpolation.html

The most important of those three articles is the third on--specifically the section on "server reconciliation," however reading and understanding the first two can be seen as a prerequisite as the context is necessary to understand the third one.

I believe that if you read and understand all three of those articles you should have sufficient knowledge to understand this PR and you should be able to easily see why this change would be useful.

To clarify, yes, the approach that @akien-mga suggested is possible, however:
a) looking at the [implementation of `KinematicBody::move_and_slide`](https://github.com/godotengine/godot/blob/master/scene/3d/physics_body_3d.cpp#L945) it's not exactly what I would call straightforward or trivial
b) This enhancement _is_ trivial as well as backwards compatible, so I believe that including it has no downsides.

Of course, the absolute ideal scenario is if `move_and_slide` didn't inject the physics delta to begin with, but that's a backwards incompatible change at this point, so this is a fair compromise imo.

**Comment by jonbonazza** [comment]
I also want to point out that because this does not exist, @AndreaCatania 's PR for the NetworkController node needs to do some hacky stuff to synchronize deltas for his use case.

**Comment by jonbonazza** [comment]
@clayjohn @akien-mga @madmiraal was that explanation sufficient?

**Comment by madmiraal** [comment]
@jonbonazza I have read the Gabriel Gambetta articles, and they highlight important considerations, but this PR does not implement them, and nor do I think it's required to implement them.

You should consider what @akien-mga suggested:
> There should be a proposal on godot-proposals that describes the use case, potential workarounds, etc., so that core developers can assess whether this is a good change to the API.
>
> One way to demonstrate the validity of this change could be to make a demo project using the feature for something that cannot be done in the engine currently.

**Comment by jonbonazza** [comment]
@madmiraal, the 3rd article i linked explicitly explains why this change is necessary. If you still can't grasp it after that, i'm not sure what else I can do to get you to understand. This PR is not to implement the things in that article, it's to allow for the strategies in that article to be implemented in all scenarios.

Ill try one more time to explain.

Many games slow down the physics tick rate on the server (e.g. 10fps) to conserve on cpu and bandwidth, while the client operates at a higher physics tock rate (i.e. 60fps). So server state is only sent to the clients every 0.1s. when the client goes to move the character, it performs prediction using the client's physics delta (which in this e.g is 1/60). Then, every physics frame, it's running its reconciliation step, where it resets the current transform to that of the last known server position (remember, this is sent to clients ever 0.1s) and replays all inputs locally since that server state was received. Since the move_and_slide calls are issued with different physics deltas on the server and the client, the client state after prediction and the state after aerver reconciliation won't match, resulting in visual jitter seen on the player's client. 

As @akien mentioned, this could be resolved by reimplementing move_and_slide manually, however that is not trivial. This change is entire backwards compatible, unintrusive, and trivial. It provides a convenient way to solve this problem by simply sending the client physics fps to the server and the server overrides the physics fps in its own calls to move_and_slide. 

I find it a little unfair to me to suggest that i trash this pr and create a proposal when i have been trying for 8 months to get the right eyes on this pr--long before the proposal process was implemented and no one even took the time to reply letting me know what next steps were.

Furthermore, i do already have a couple proposals open, and have been vigilantly monitoring the proposals repo. I dont see anyone from the dev team giving any feedback to those proposals and there doesn't seem to be a clear process for next steps after a proposal is created. If prs are being rejected because there is no proposal, yet proposals are being ignored, what means left is there to get a change in? This change is only a handful of lines at that. I can't even fathom what kind of beurocracy hoes into complex prs!

It should also be mentioned that as

**Comment by reduz** [comment]
@jonbonazza Apologies for not getting involved earlier. Unfortunately 4.0 is a huge release that requires too much of my time, so I am not usually as available to help other contributors. I hope this situation reverts after its out, and I can again dedicate more of my time to review PRs and help contributors.

The reason why I am against merging this PR is explained in [point #3 of the Best Practices for Engine Contributors](https://docs.godotengine.org/en/stable/community/contributing/best_practices_for_engine_contributors.html#the-problem-has-to-be-complex-or-frequent) page.

This feature fails to meet either the complexity or frequency criteria. Re-implementing your own move_and_slide that fits your needs better is not a task of high difficulty. move_and_slide is intended to be a wrapper over move_and_collide for the most common use cases, and this is clearly not a common use case, but it's not very complex. Its less than 100 lines of code you can easily re-implement in GDScript, C# or anything that meets your needs. If you can implement a client-server physics architecture, I fail to see how the complexity of this task can be above your head.

move_and_slide is already pretty bloated as-is, which means we will be rewriting how it works for 4.0 with an aim towards making it even simpler to use (yet still covering the _common_ use cases). I would rather not touch it for the time being.

**Comment by reduz** [comment]
One thing we could do, is include an example of writing a move_and_slide function in GDScript, that users can take and customize if they really want to.

**Comment by jonbonazza** [comment]
Hi @reduz ,
Thanks for taking a look.

If move_and_slide is being rewritten for 4.0 anyway then it's probably not even worth arguing at this point. I will request that when you fo revisit the function for 4.0, you don't automatically include the physics process interval and just leave it up to the user to decide whether they want to include it.

I suppose this can be closed now.

**Pull request closed without merging**
